doc/man/quickstart.5.md
% quickstart(5)
% Andrew Gaffney, Lucas Ramage
% October 16, 2006

# NAME

quickstart - profile configuration file.

# DESCRIPTION

The follow is a list of all the valid configuration options with description
and valid parameters.

**install_mode**

:   choose the install mode.

Usage:

  install_mode [mode]

Parameters:

  mode - the install mode (normal, chroot, or stage4). A normal install is for
         an install from a stage3 with bootloader setup. A chroot install is
	 the same as a normal install, except it skips the partitioning and
	 bootloader setup. A stage4 install is like a normal install, except
	 that it skips straight from unpacking the stage tarball to configuring
	 the bootloader.

**part**

:   create a partition.

Usage:

  part [drive] [minor] [type] [size]

Parameters:

  drive - the drive to add this partition e.g., `hda`, `sdb`, etc.

  minor - the partition number; these should be in order.

  type - the partition type used in fdisk e.g., `82`, `83`, or `"extended"`.

  size - the size of the partition e.g., `32MB` or `15GB`, whole numbers only,
         `+` for remaining space, or `n%` for percentage of remaining
	 (these should only be used after all specific partition sizes are
	 specified), leave blank for "extended".

**mdraid**

:   create a Mirrored Disk (MD) RAID array.

Parameters:

  array - name of the array e.g., `md0`, `md1`, etc.

  arrayopts - arguments to `mdadm` after `--create /dev/mdX` e.g.,
              `-l 1 -n 2 /dev/sda2 /dev/sdb2`.

**lvm_volgroup**

:   create an LVM volume group.

Usage:

  lvm_volgroup [name] [devices]

Parameters:

  name - name of the volume group to create.

  devices - list of block devices to include in the volume group.

**lvm_logvol**

:   create an LVM logical volume.

Usage:

  lvm_logvol [volgroup] [size] [name]

Parameters:

  volgroup - name of a volume group created with `lvm_volgroup`.

  size - size of logical volume to pass to `lvcreate`.

  name - name of logical volume to pass to `lvcreate`.

**format**

:   format a partition.

Usage:

  format [device] [fs]

Parameters:

  device - the device to format e.g., `/dev/hda2` or `/dev/sdb4`.

  fs - the filesystem to use e.g., `ext2`, `ext3`, or `swap`.

**mountfs**

:   mount a filesystem.

Usage:

  mountfs [device] [type] [mountpoint] [mountopts]

Parameters:

  device - the device to mount e.g., `/dev/hda2` or `/dev/sdb4`.

  type - filesystem of device (use `auto` if you're not sure).

  mountpoint - the directory to mount the filesystem.

  mountopts - additional options to use when mounting.

**netmount**

:   mount a network filesystem.

Usage:

  netmount [export] [type] [mountpoint] [mountopts]

Parameters:

  export - path to the network filesystem e.g., `192.168.0.12:/usr/portage`.

  type - network filesystem type e.g., `nfs`, `smbfs`, `cifs`, etc.

  mountpoint - the directory to mount the filesystem.

  mountopts - additional options to use when mounting.

**bootloader**

:   specify the bootloader to use (defaults to `grub`).

Usage:

  bootloader [bootloader]

Parameters:

  bootloader - the name of the bootloader to use e.g., `grub` or `lilo`.

**logger**

:   specify the bootloader to use (defaults to `syslog-ng`).

Usage:

  logger [logger]

Parameters:

  logger - the name of the logger to use e.g., `metalog` or `syslog-ng`.

**cron**

:   specify the cron daemon to use (defaults to `vixie-cron`).

Usage:

  cron [cron]

Parameters:

  cron - the name of the cron daemon to use e.g., `vixie-cron`.
         one can also specify `none` here to skip installing a cron daemon.

**rootpw**

:   set the root password (required if not using `rootpw_crypt`).

Usage:

  rootpw [password]

Parameters:

  password - the plain-text password for the root user.

**rootpw_crypt**

:   set the root password (required if not using `rootpw`).

Usage:

  rootpw_crypt [password]

Parameters:

  password - the pre-hashed password for the root user.

**stage_uri**

:   specify the URI to the stage tarball (required).

Usage:

  stage_uri [uri]

Parameters:

  uri - URI to the location of the stage tarball.
        Supported protocols include: `http`, `https`, `ftp`, or `file`.

**tree_type**

:   specify the portage tree type (defaults to sync)

Usage:

  tree_type [type] [uri]

Parameters:

  type - the method for fetching the portage tree (`sync`, `webrsync`, or
         `snapshot`). One can also specify `none` to skip this step.

  uri - URI to the location of the portage snapshot if `snapshot` was
        specified for the type.

**bootloader_install_device**

:   specify the device to install the bootloader to (defaults to MBR of device
    `/boot` is on).

Usage:

  bootloader_install_device [device]

Parameters:

  device - device to install bootloader to e.g., `/dev/hda`, `/dev/hda3`, etc.

**chroot_dir**

:   specify the directory to use for the `chroot` (defaults to `/mnt/gentoo`).

Usage:

  chroot_dir [dir]

Parameters:

  dir - directory to use for the `chroot`.

**extra_packages**

:   specify extra packages to emerge.

Usage:

  extra_packages [pkg]

Parameters:

  pkg - package (or packages, space-separated) to emerge.

**genkernel_opts**

:   specify extra options to pass to `genkernel`.

Usage:

  genkernel_opts [opts]

Parameters:

  opts - the extra options to pass to `genkernel`.

**kernel_config_uri**

:   specify the URI to a custom kernel config.

Usage:

kernel_config_uri [uri]

Parameters:

  uri - URI to the location of the custom kernel config.

**kernel_sources**

:   specify the kernel sources to use (defaults to `gentoo-sources`).

Usage:

  kernel_sources [source]

Parameters:

  source - kernel sources to emerge.

**bootloader_kernel_args**

:   specify extra commandline arguments to pass to the kernel.

Usage:

  bootloader_kernel_args [args]

Parameters:

  args - extra commandline arguments to pass to the kernel.

**timezone**

:   specify the timezone.

Usage:

timezone [tz]

Parameters:

  tz - timezone to use (relative to `/usr/share/zoneinfo/`).

  rcadd - add the specified service to the specified runlevel.

Usage:

  rcadd [service] [runlevel]

Parameters:

  service - name of service to add.

  runlevel - runlevel to add service to.

  rcdel - remove the specified service from the specified runlevel.

Usage:

  rcdel [service] [runlevel]

Parameters:

  service - name of service to remove.

  runlevel - runlevel to remove service from (usually not required).

**net**

:   setup networking.

Usage:

  net [device] [ip] [gateway]

Parameters:

  device  - network device e.g., `eth0`.

  ip - static IP address, `"dhcp"` or `"current"`, which tries to use the
       current configuration.

  gateway - gateway IP if using a static IP.

**skip**

:   skips an install step.

Usage:

  skip [step]

Parameters:

  step - name of install step to skip.

**server**

:   specify a quickstart server.

Usage:

  server [host]:[port]

Parameters:

  host - hostname or IP address of quickstart server.

  port - port server is listening on (defaults to `8899`).

**use_linux32**

:   enable the use of linux32 for doing 32ul installs on 64-bit boxes.

Usage:

  use_linux32

**verbose**

:   enable verbose mode (same as `--verbose`).

Usage:

  verbose

# EXAMPLE

```sh
stage_uri http://192.168.0.12/gentoo/stage3-i686-2006.1.tar.bz2
tree_type snapshot http://192.168.0.12/gentoo/portage-20061216.tar.bz2
rootpw password
bootloader grub

part hda 1 83 100M
part hda 2 82 512M
part hda 3 83 +

format /dev/hda1 ext2
format /dev/hda2 swap
format /dev/hda3 ext3

mountfs /dev/hda1 ext2 /boot
mountfs /dev/hda2 swap
mountfs /dev/hda3 ext3 / noatime

net eth0 dhcp

netmount 192.168.0.12:/usr/portage nfs /usr/portage ro

post_install_portage_tree() {
  cat > ${chroot_dir}/etc/make.conf <<EOF
CHOST="i686-pc-linux-gnu"
CFLAGS="-O2 -march=athlon-xp -pipe"
CXXFLAGS="\${CFLAGS}"
USE="-X -gtk -gnome -kde -qt"
EOF

  echo "portdbapi.auxdbmodule = cache.metadata_overlay.database" \
    > ${chroot_dir}/etc/portage/modules
}
```



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [12.7.0] - 2012-07-07

### Added

- Partitioning from a blank disk (drive is wiped).
- Specify partition sizes in MB, GB, % of remaining, or + (all remaining).
- Creation of md raid arrays.
- Creation of lvm2 volumes.
- Ability to format partitions as ext2, ext3, swap, reiserfs, xfs, or jfs.
- Specify local filesystems to be mounted during the install.
- Specify network shares to be mounted during the install.
- Choose your bootloader.
- Automatic bootloader configuration (currently grub, palo, and silo).
- Specify device for bootloader installation.
- Choose your logger (or none).
- Choose your cron daemon (or none).
- Choose your root password (plain-text or pre-encrypted).
- Specify URI for stage 3 tarball (file, http, https, ftp, or rsync).
- Specify method for getting a portage tree (sync, webrsync, or snapshot).
- Specify the directory that is used for the chroot.
- Specify extra packages to be emerged after the base system.
- Specify extra options passed to genkernel.
- Specify URI for pre-made kernel config.
- Choose which kernel sources package to use to build your kernel.
- Choose your timezone.
- Choose which services to add to which runlevels.
- Choose which services to remove from which runlevels.
- Specify basic networking configuration.
- Support for custom code using pre-/post-install step hooks in the config file.  

### Changed

- Config syntax modeled on the [Kickstart](https://en.wikipedia.org/wiki/Kickstart_(Linux)) config syntax.

[unreleased]: https://github.com/oxr463/quickstart/compare/v12.7.0...HEAD
[12.7.0]: https://github.com/oxr463/quickstart/releases/tag/v12.7.0


src/quickstart.sh
#!/bin/sh
# An installer for Gentoo Linux written in POSIX shell
# SPDX-License-Identifier: GPL-2.0-only
set -eu

# Constants
VERSION="12.7.1"

# Options vars
debug=0
verbose=0
quiet=0
sanitycheck=0

import() {
  module=$1

  # shellcheck source=/dev/null
  . "modules/${module}.sh"
  debug import "imported module ${module}"
}

usage() {
  msg=$1

  if [ -n "${msg}" ]; then
    printf "%s\n" "${msg}"
  fi
  cat <<EOF
Usage:
  quickstart [-h|--help] [-d|--debug] [-v|--verbose] [-q|--quiet]
             [-s|--sanity-check] [--version] <profile>

Options:
  -h|--help            Show this message and quit
  -d|--debug           Output debugging messages
  -q|--quiet           Only output fatal error messages
  -v|--verbose         Be verbose (show external command output)
  -s|--sanity-check    Sanity check install profile and exit
  -c|--client <host>   Act as a client and connect to a quickstartd
  --version            Print version and exit

Arguments:
  profile              Path to an install profile
EOF
}

# Import modules
import output
import util
import spawn
import fetch
import portage
import bootloader
import partition
import step
import config
import stepcontrol
import server

# Parse args
while [ ${#} -gt 0 ]
do
  a=${1}
  shift
  case "${a}" in
    -h|--help)
      usage
      exit 0
      ;;
    -s|--sanity-check)
      sanitycheck=1
      ;;
    -d|--debug)
      # shellcheck disable=SC2034
      debug=1
      ;;
    -q|--quiet)
      if [ ${verbose} = 1 ]; then
        usage "The --quiet and --verbose options are mutually exclusive"
        exit 1
      fi
      quiet=1
      ;;
    -v|--verbose)
      if [ ${quiet} = 1 ]; then
        usage "The --quiet and --verbose options are mutually exclusive"
        exit 1
      fi
      verbose=1
      ;;
    -c|--client)
      server=${1}
      shift
      ;;
    --version)
      echo "Quickstart version ${VERSION}"
      exit 0
      ;;
    -*)
      usage "You have specified an invalid option: ${a}"
      exit 1
      ;;
    *)
      profile=$a
      ;;
  esac
done

if [ -n "${server}" ]; then
  server_init
  if server_get_profile; then
    profile="/tmp/quickstart_profile"
  fi
fi

if [ -z "${profile}" ]; then
  usage "You must specify a profile"
  exit 1
fi

if [ ! -f "${profile}" ]; then
  error "Specified profile does not exist!"
  exit 1
else
  # shellcheck source=/dev/null
  . "${profile}"
  if ! touch "${logfile:?}" 2>/dev/null; then
    error "Logfile is not writeable!"
    exit 1
  fi
  arch=$(get_arch)
  debug main "arch is ${arch}"
  [ -z "${arch}" ] && die "Could not determine arch!"
  runstep sanity_check_config "Sanity checking config"
  if [ "${sanitycheck}" = "1" ]; then
    debug main "Exiting due to --sanity-check"
    exit
  fi
fi


[ -z "${install_mode}" ] && install_mode="normal"

run_pre_install_script "Running pre-install script"

if [ "${install_mode}" != "chroot" ]; then 
  runstep partition "Partitioning"
fi

runstep setup_md_raid "Setting up RAID arrays"
runstep setup_lvm "Setting up LVM volumes"
runstep format_devices "Formatting devices"
runstep mount_local_partitions "Mounting partitions"
runstep mount_network_shares "Mounting network shares"
runstep unpack_stage_tarball "Fetching and unpacking stage tarball"
runstep prepare_chroot "Preparing chroot"

if [ "${install_mode}" != "stage4" ]; then
  runstep install_portage_tree "Installing portage tree"
  runstep set_root_password "Setting root password"
  runstep set_timezone "Setting timezone"
  runstep setup_fstab "Setting up /etc/fstab"
  runstep build_kernel "Building kernel"
  runstep install_logging_daemon "Installing logging daemon"
  runstep install_cron_daemon "Installing cron daemon"
  runstep setup_network_post "Setting up post-install networking"
  runstep install_bootloader "Installing bootloader"
fi

if [ "${install_mode}" != "chroot" ]; then
  runstep configure_bootloader "Configuring bootloader"
fi

runstep install_extra_packages "Installing extra packages"
runstep add_and_remove_services "Adding and removing services"
runstep run_post_install_script "Running post-install script"
runstep finishing_cleanup "Cleaning up"

notify "Install complete!"

if [ "${reboot:=no}" = "yes" ]; then
  notify "Rebooting..."
  reboot
fi
